### PR TITLE
support tornado-6.x, support coroutine message_handler

### PR DESCRIPTION
**Pull request by ploxiln**
~~needs a couple tests for the new coroutine support~~ (done)
needs pydoc updates

could refactor a bit to use coroutines in one or two more places internally ... tests with mocks make this annoying

### CODE REVIEW

**Comment by Integralist** [comment]
😍

**Comment by Integralist** [comment]
Can I enquire as to the process for this PR getting reviewed/approved? What steps are outstanding? 

I'm eager to see Tornado 6 supported but am unclear on how long these things typically take to get over the finish line (I appreciate this is unlikely someone's day job and so that will add to the over time frame).

Thanks!

**Comment by ploxiln** [comment]
I pushed this up about a week before I went on a planned vacation, and I'm not quite home yet ... so it'll probably be another 2 weeks before I add some new tests and give this a sanity-test in a real environment, and then hopefully some others have some time for review ... you can see a typical timeline in the tornado-5 compat PR: https://github.com/nsqio/pynsq/pull/232

**Comment by katzdm** [comment]
Hey @ploxiln , just wanted to ping this thread. Really excited for this to land; thanks for moving this forward!

**Comment by Integralist** [comment]
Hello @ploxiln and other friends! 

Just wondering if there is anything that people (such as myself) outside of the NSQ organisation can do to help test and get this PR over the finish line? ❤️

**Comment by ploxiln** [comment]
I'm still not sure when this will be merged, but you can start testing it now in your own applications if you like - I added the (minimal) coroutine tests I intended, it seems to work fine, in my real-world testing.

**Comment by Integralist** [comment]
Hi @ploxiln 

Would it be possible to tag this as a beta release? It makes testing for us easier and we've used that approach for rolling out services into production, for example, using `pynsq==0.9.0-beta1` in the past.

Also, with regards to testing, we have quite a few services that we would like to migrate to the latest tornado version and so it would be good to understand what feedback you require in order to consider this PR mergeable (e.g. what is preventing this PR from being merged at this point? other than some integrated testing).

Thanks!

**Comment by mreiferson** [comment]
(this needs a rebase now that the GitHub Actions PR landed)

I guess I'll add this to my list of things I need to take a look at, but would still be useful to hear if y'all actually got this up and running on any canary services @Integralist and @katzdm?

File: nsq/reader.py [hunk 1]
```diff
@@ -313,27 +314,43 @@ def _on_message(self, conn, message, **kwargs):
...
                 return message.requeue()
 
-        if not message.is_async() and not message.has_responded():
-            assert success is not None, 'ambiguous return value for synchronous mode'
-            if success:
+        if result not in (True, False, None):
+            # assume handler returned a Future or Coroutine
+            message.enable_async()
+            fut = tornado.gen.convert_yielded(result)
+            fut.add_done_callback(functools.partial(self._maybe_finish, message))
```
Comment (mreiferson) [inline nsq/reader.py:336]: This is a subtle addition, but I like it! You can now implicitly enable async by returning a future or coroutine.

Probably something we should document?

Comment (ploxiln) [reply nsq/reader.py:336]: yeah ... but the callback attached to the future will call `message.finish()` automatically, so this acts kinda like a synchronous handler function ... I'll see about updating the docs

Comment (mreiferson) [reply nsq/reader.py:336]: ahh, right, nevermind

**Comment by Integralist** [comment]
@mreiferson @ploxiln just a 'heads up' that I'm letting this code soak on one of our environments and will let you know how things look in the next day.

**Comment by Integralist** [comment]
@mreiferson been soaking the past day and it looks fine to me. 👍 

I think we're going to get our Data Infra team to do some testing also. So maybe best to wait for a 👍 from them shortly (unless you're willing to trust my approval)

**Comment by ploxiln** [comment]
Thank you @Integralist for trying this out. (I've also been running this for over a month now,  but not really exercising 100% of it.)

I pushed up some minor docs changes, if they look good I'll squash them in.

@mreiferson I think we could release a 0.9.0b2 after merging this, to give it more easier testing before stable release. But it would probably be fine to release 0.9.0 final with this, and then 0.9.1 with fixes for discovered bugs.

**Comment by Integralist** [comment]
Thanks @ploxiln the first doc change looks fine. I was a bit confused by the tornado 5.x reference considering these changes are for supporting tornado 6.x so I presume the 'docs' have a unrelated dependency on tornado?

**Comment by ploxiln** [comment]
Correct, that's just for references to tornado types when loading pynsq source files, in order to extract doc comments from pynsq classes. Might as well keep compatibility with python-2.7 for generating docs for now.

**Comment by mreiferson** [comment]
I pushed https://pypi.org/manage/project/pynsq/release/0.9.0b2/

We haven't traditionally _tagged_ (git) pre-releases, so `v0.9.0b1` was a bit of an anomaly.

**Pull request merged**